### PR TITLE
Override munit ExecutionContext with one from IORuntime

### PR DESCRIPTION
**Pull request by milanvdm**
Closes https://github.com/typelevel/munit-cats-effect/issues/110

This overrides the default munit `parasitic-execution-context` with the one from `IORuntime.global`. 

Not sure if tests can be added for ScalaJS on the issue that @djspiewak mentions as I have no ScalaJS knowledge.

### CODE REVIEW

**Comment by milanvdm** [comment]
The tests seem to be flaky. I will investigate further tomorrow as there is no clear test failing as far as I can tell.

**Comment by milanvdm** [comment]
This PR is currently blocked by a bug in cats-effect in the blocking part of the compute pool. I Will follow up once fixed.

File: ce3/shared/src/main/scala/munit/CatsEffectSuite.scala [hunk 1]
```diff
@@ -16,18 +16,20 @@
...
+import scala.concurrent.{ExecutionContext, Future}
 
 abstract class CatsEffectSuite
     extends FunSuite
     with CatsEffectAssertions
     with CatsEffectFixtures
     with CatsEffectFunFixtures {
 
-  implicit val ioRuntime: IORuntime = IORuntime.global
+  implicit def munitIoRuntime: IORuntime = IORuntime.global
```
Comment (danicheg) [inline ce3/shared/src/main/scala/munit/CatsEffectSuite.scala:30]: I don't get why was `val` changed on `def` here. `IORuntime.global` is a `val` though.

**Review by danicheg — APPROVED** [review]
LGTM. The question is for my understanding rather than any block on this.

**Review by mpilquist — APPROVED** [review]

File: ce3/shared/src/main/scala/munit/CatsEffectSuite.scala (hunk 1 above)
Comment (milanvdm) [reply ce3/shared/src/main/scala/munit/CatsEffectSuite.scala:30]: When offering an interface to a user, they might want to overwrite this `munitIoRuntime` with a `def`.

The same reason that munit itself has the `munitExecutionContext` as a `def` on the class we are overwriting.

**Comment by vasilmkd** [comment]
@milanvdm @armanbilge Maybe it is best to revert this change and wait until Cats Effect `3.3` releases to incorporate it.

**Comment by vasilmkd** [comment]
On the other hand, maybe it is not wise to have this change on the JVM at all. Every time a `Future` is awaited on the main pool, it would send a signal to the pool to start new threads. And in huge test suites, that could result in thousands of threads being created. I don't know what exactly to think.

**Comment by vasilmkd** [comment]
Please also don't rush to make changes, we'll continue to monitor the possible implications of this PR.

**Comment by armanbilge** [comment]
> every time a Future is awaited on the compute pool, it would send a signal to the pool to start new threads. And in huge test suites, that could result in thousands of threads being created.

IIUC from our Discord discussion, there is (at least) one `Future` created per test, when the `IO` is converted.

The [munit docs](https://scalameta.org/munit/docs/tests.html#run-tests-in-parallel) say that
> MUnit does not support running individual test cases in parallel. However, sbt automatically parallelizes the execution of multiple test suites. 

So, won't the number of these per-test `Future`s be bounded by the overall parallelism level that sbt is using?

**Comment by vasilmkd** [comment]
I published `cats-effect` and `munit-cats-effect-3` local versions which print when new threads are started, and this seems to not spawn anything in `coreJVM/test` on `fs2`, so we're good. I guess I underestimated how these things work. Let's keep it for now.

**Comment by ahjohannessen** [comment]
This change seems to fail fs2-grpc server streaming tests with regards to [cancellation](https://github.com/typelevel/fs2-grpc/runs/3850697415?check_suite_focus=true#step:6:77). Not sure what is happening, but restoring the previous execution context in `Fs2GrpcSuite` makes tests pass again:

```scala
  private val parasiticExecutionContext = new ExecutionContext {
    def execute(runnable: Runnable): Unit = runnable.run()
    def reportFailure(cause: Throwable): Unit = cause.printStackTrace()
  }

  override val munitExecutionContext: ExecutionContext = parasiticExecutionContext
  ```

**Pull request merged**